### ISSUE TITLE
Figure out how to handle lengths and percentages in transform animations with translations

### ISSUE THREAD

**Issue by Manishearth** [servo/servo#13267]
We currently just implement `Interpolate` on transforms. Most of the transform functions just operate on floats, and the one which doesn't (`translate`) is simple to interpolate sicne you can just interpolate the contained calcs.

However, the spec allows you to interpolate between different transform functions. It does this via a set of steps being implementing in #﻿13188, which decompose the from and to matrices into their component operations, interpolate those, and compose them again.

This works fine for cases implemented in #13188 (where you have the same from and to functions -- the matrix decomposition still needs to be done for `matrix()`, `rotate()` and `perspective()`), but if you need to interpolate between `translate(10px, 20%)` and `rotate(45deg)`, you need to convert the `10px` and `20%` into a translation matrix somehow, which you can't do without the reference box. calcs can be interpolated with other calcs, but they can't be a part of a more complicated algorithm like the one here.

We could just pass the reference box down in interpolation, or special-case these. (The other advantage of special casing is that we can cache the decomposed values on the stack somewhere)

What do you think?

cc @SimonSapin @emilio

**Comment by Manishearth** [comment]
Not entirely sure how gecko handles this, seems like it synthesizes a fake "interpolatematrix" transform function and passes it on to layout which interpolates it with the ref box.

**Comment by emilio** [comment]
Hmm... How are we supposed to have the reference box from code like, let's say, animations?

**Comment by emilio** [comment]
Hmmm... I wonder if we could have a computed value which means "interpolated list of matri

**Comment by emilio** [comment]
So the magic in Gecko happens on display list construction, when you already have the frame laid out (and hence the full reference box). In the style system they're stored as a list of `nsCSSValue`s...

searchfox.org/mozilla-central/source/layout/base/nsDisplayList.cpp#5755

**Comment by emilio** [comment]
Sorry, I meant to link to `AddAnimationsForProperty`, that gets the animations for the compositor.

But yeah, it seems that they store the computed value of a transform as a list of `nsCSSValue` arrays (which would be equivalent to our `TransformOperation`s, that in the case of `interpolatematrix` contains two transform lists, that they read separately, then interpolate.

In Servo, the relevant code would be `create_stacking_context` in `display_list_builder.rs`, where we also have the computed border box of the fragment... It seems that we'd need to postpone the interpolation until there, storing in the computed value a TransformOperation that contains the transforms to interpolate (I *think*).

But yeah, it seems this needs pretty much a special case, @Manishearth.

**Comment by Manishearth** [comment]
The only time the refbox is required is when you're interpolating between lists of different functions where at least one of the lists contains a translate.

Think we should add the interpolatematrix function only for that case, or just move all decomposition animation to the display list?

**Comment by emilio** [comment]
I think it makes sense for all the interpolation code to be together.

**Comment by BorisChiou** [comment]
In order to resolve this issue, we have to finish the implementation of converting InterpolateMatrix and AccumulateMatrix into a transform matrix [1].

[1] https://github.com/servo/servo/blob/2a97b2cd7cb7e2e52ea48f857d1687b8dcee6604/components/style/values/computed/transform.rs#L121-L128

**Comment by mrobinson** [comment]
Is it possible that this complete now?

**Comment by Manishearth** [comment]
I don't think so, we have CSS-level support for this but not layout support (the `to_3d_matrix` function). We're close, though, it's just that one part that needs to be implemented.
